### PR TITLE
[ObjectFifo] Add pass to flatten the logical objectFifo

### PR DESCRIPTION
**Pull request by yzhang93**
This PR is part to achieve https://github.com/nod-ai/iree-amd-aie/issues/644

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFlattenLogicalObjectFifo.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    // operations converted into objectfifos can have a different source and
+    // target type.
+    for (Operation *user : newLogicalObjectFifo->getUsers()) {
+      if (auto accessOp = dyn_cast<AMDAIE::LogicalObjectFifoAccessOp>(user)) {
+        rewriter.setInsertionPoint(accessOp);
+        auto newAccessOp = rewriter.create<AMDAIE::LogicalObjectFifoAccessOp>(
+            rewriter.getUnknownLoc(), newLogicalObjectFifo.getOutput(),
+            accessOp.getAccessType());
+
+        llvm::ArrayRef<int64_t> sizes = oldType.getShape();
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFlattenLogicalObjectFifo.cpp:76]: No need to store `oldType.getShape()` if there are no other uses other than the one in `rewriter.create<memref::ReinterpretCastOp>` below.

**Review by Abhishek-Varma — CHANGES_REQUESTED** [review]
Few comments.

Could you come up with a smaller lit test?
Essentially the pass is to demo the linearization of the memref involved in `logicalobjectfifo.from_memref` + `logicalobjectfifo.access`. So, we should be able to form a smaller lit test.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFlattenLogicalObjectFifo.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    if (castOp) {
+      oldType = cast<MemRefType>(castOp.getSource().getType());
+    }
+
+    // Get linearized size and new type accordingly.
+    SmallVector<int64_t> shape = llvm::to_vector(oldType.getShape());
+    int64_t linearizedSize = 1;
+    for (auto s : shape) {
+      linearizedSize *= s;
+    }
```
[the remark this answers is not in the document]
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFlattenLogicalObjectFifo.cpp]: Good point, thanks!

**Review by Abhishek-Varma — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**